### PR TITLE
win_chocolatey: refactor module to fix bugs and add new features

### LINKED ISSUE #24569: Install Chocolatey from Internal Source
Originally at https://github.com/ansible/ansible-modules-extras/issues/2916 @dagwieers suggested it get moved here.

cc @nitzmahone 
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME
win_chocolatey

##### OS / ENVIRONMENT
Windows

##### SUMMARY
Folks would like to be able to have Chocolatey installed from internal sources when using the Ansible module, particularly when they are completely offline.

This would be a good capability added to the module. It is already supported by other config mgmt tools so those could provide good references on how to add it to the chocolatey ansible module.

### LINKED ISSUE #30784: win_chocolatey, support source username and password
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
win_chocolatey

##### SUMMARY
Chocolatey supports sources that require authentication. The module `win_chocolatey` currently doesn't have a way to pass `--user` and `--password` [options](https://github.com/chocolatey/choco/wiki/CommandsInstall#options-and-switches) to `choco install`.

### LINKED ISSUE #42026: Add example of how to install Chocolatey from internal source
##### SUMMARY
The win_chocolatey module installs Chocolatey from a public URL if it is not present. This can cause issues for people who are in a locked down environment. While we can add support for specifying a different URL for the Chocolatey install we have decided to go a different path and recommend people install Chocolatey through other tasks like win_shell.

While having win_chocolatey install Chocolatey itself is good for demo's and getting started, we don't want to add any more logic for this because of the following reasons;

* We would start to re-implement the wheel for various win_get_url options like authorisation, force basic auth, timeouts which is not right
* There can be a disconnect between the source/version args when we need to bootstrap Chocolatey automatically but the user specified a different package in the name
* This would expose win_chocolatey to running a potentially arbitrary script from a URL, this is still an issue today though

This could potentially still be done with win_chocolatey in the future but we need to play around with the various options and move on from there once we know more.

What we need to do is create some examples in the win_chocolatey on how to install Chocolatey from a url or file path.

Slight continuation of https://github.com/ansible/ansible/issues/24569

##### ISSUE TYPE
 - Documentation Report

##### COMPONENT NAME
win_chocolatey

##### ANSIBLE VERSION
```
devel
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
Chocolatey on Windows hosts

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
As part of my Chocolatey craze this is the final PR to refactor the existing win_chocolatey module. This refactor does the following

* Implement source username/password
    * Supercedes https://github.com/ansible/ansible/pull/41479 
    * Fixes https://github.com/ansible/ansible/issues/30784
* Ability to specify a source when installing Chocolatey itself
    * Fixes https://github.com/ansible/ansible/issues/24569
* Added an example on the above
    * Fixes https://github.com/ansible/ansible/issues/42026
* Fixed an issue where `state: downgrade` would actually upgrade a package if no version was set
* Fixed issue where choco was upgrade in check mode if the version did not meet the requirements
* Simplify the proxy user/pass to no longer require `"` to be escaped (added to porting guide as well)
* Expanded the docs to cover more information around the parameters and added some more examples
* Added some references to the new source/config/feature Chocolatey modules
* Added the ability to specify multiple packages in 1 module invocation, saving on the time it takes to install multiple packages
* Expanded the tests to cover a lot of the scenarios that win_chocolatey can be used for and are now less reliant on a stable internet connection

~~I still need to manually test some of the Chocolatey bootstrapping tests before merging this in. Unfortunately I can't find a good way to cover these tests under CI so it will be manually done for now.~~

Update: have tested the latest changes against a choco server implementation and verified it works with and without basic auth.

##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
win_chocolatey

##### ANSIBLE VERSION
```
2.7
```

### CODE REVIEW

**Comment by jborean93** [comment]
cc @gep13 @ferventcoder

File: lib/ansible/modules/windows/win_chocolatey.ps1 [hunk 1]
```diff
@@ -16,532 +19,543 @@ $params = Parse-Args $args -supports_check_mode $true
...
-            $script:options = @( "-dv" )
-            Choco-Upgrade -package chocolatey  -proxy_url $proxy_url -proxy_username $proxy_username -proxy_password $proxy_password
-        }
+Function Get-CommonChocolateyArguments {
+    # uses global vars like check_mode and verbosity to control the common args
+    # run with Chocolatey
+    $arguments = [System.Collections.ArrayList]@("--yes", "--no-progress")
+    # global vars that control the arguments
+    if ($check_mode) {
+        $arguments.Add("--what-if") > $null
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.ps1:59]: I am surprised to see `> $null` here.
Why is this needed for something basic as adding an item to a list ?

File: lib/ansible/modules/windows/win_chocolatey.ps1 [hunk 2]
```diff
@@ -16,532 +19,543 @@ $params = Parse-Args $args -supports_check_mode $true
...
-                -ignorechecksums $ignorechecksums -ignoredependencies $ignoredependencies `
-                -allowdowngrade $allowdowngrade -proxy_url $proxy_url `
-                -proxy_username $proxy_username -proxy_password $proxy_password `
-                -allowprerelease $allowprerelease -architecture $architecture
-            return
+    $choco_app = Get-Command -Name choco.exe -CommandType Application -ErrorAction SilentlyContinue
+    if ($null -eq $choco_app) {
+        # We need to install chocolatey
+        # Enable TLS1.1/TLS1.2 if they're available but disabled (eg. .NET 4.5)
+        $security_protcols = [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::SystemDefault
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.ps1]: I fixed this in another PR, but you have a typo in `security_protcols`.

File: lib/ansible/modules/windows/win_chocolatey.ps1 [hunk 3]
```diff
@@ -16,532 +19,543 @@ $params = Parse-Args $args -supports_check_mode $true
...
+                # http://host/install.ps1, we rely on this behaviour when a
+                # user specifies the choco source URL. If a custom URL or file
+                # path is desired, they should use win_get_url/win_shell
+                # manually
+                # we need to strip the path off the URL and append install.ps1
+                $uri_info = [System.Uri]$source
+                $script_url = "$($uri_info.Scheme)/$($uri_info.Authority)/install.ps1"
+            }
+        } else {
+            $script_url = "https://chocolatey.org/install.ps1"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.ps1]: Could we not have the `source` value default to 'https://chocolatey.org/install.ps1' ?
At least it would allow us to explicitly do: `source: null` if we don't want any installation.

File: lib/ansible/modules/windows/win_chocolatey.ps1 [hunk 4]
```diff
@@ -16,532 +19,543 @@ $params = Parse-Args $args -supports_check_mode $true
...
 
-    if ($installargs)
-    {
-        $options += "--installargs", $installargs
-    }
+        $install_script = $client.DownloadString($script_url)
+        if (-not $check_mode) {
+            $res = Run-Command -command "powershell.exe -" -stdin $install_script -environment $environment
+            if ($res.rc -ne 0) {
+                Fail-Json -obj $result -message "Chocolatey bootstrap installation failed. RC: $($res.rc), STDOUT: $($res.stdout), STDERR: $($res.stderr)"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.ps1]: I would add stdout and stderr as keys in the result dictionary, as is custom with other modules.

They won't be part of the stable interface, but it would appear better (or could be accessed programmatically).

File: lib/ansible/modules/windows/win_chocolatey.ps1 [hunk 5]
```diff
@@ -16,532 +19,543 @@ $params = Parse-Args $args -supports_check_mode $true
...
+            }
+        }
+        $result.changed = $true
+        Add-Warning -obj $result -message "Chocolatey was missing from this system, so it was installed during this task run."
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.ps1]: I wonder if we have to expose the warning in check-mode. Not sure if warnings are intended to appear as if it was a proper run, especially if it seems to affirm something this explicitly.

File: lib/ansible/modules/windows/win_chocolatey.py [hunk 6]
```diff
@@ -17,111 +18,177 @@
...
+      documented by the package itself.
+    - Before Ansible 2.7, this option was just I(params).
+    type: str
+    version_added: '2.1'
+    aliases:
+    - params
   proxy_url:
     description:
-      - Proxy url used to install chocolatey and the package.
+    - Proxy url used to install chocolatey and the package.
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.py]: I would put URL in capital here. (was wrong)

File: lib/ansible/modules/windows/win_chocolatey.py [hunk 7]
```diff
@@ -166,19 +240,37 @@
...
-    architecture: 'x86'
+    architecture: x86
 
 - name: Install git from specified repository
   win_chocolatey:
     name: git
     source: https://someserver/api/v2/
 
+- name: Install git from a pre configured source (win_chocolatey_source)
+  win_chocoaltey:
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.py]: Typo

File: lib/ansible/modules/windows/win_chocolatey.py [hunk 8]
```diff
@@ -189,12 +281,8 @@
 
 - name: uninstall multiple packages
   win_chocolatey:
-    name: '{{ item }}'
+    name: procexp, putty, windirstat
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_chocolatey.py]: Should we use proper YAML in the examples? Rather than advise people to use a backward-compatible interface (this comes from k=v syntax originally).

Comment (dagwieers) [reply lib/ansible/modules/windows/win_chocolatey.py]: Also, do we need to add **win_chocolatey** to the list of modules that support squashing ?
```
# squash actions
# Ansible can optimise actions that call modules with list parameters
# when looping. Instead of calling the module once per with_ item, the
# module is called once with all items at once. Currently this only works
# under limited circumstances, and only with parameters named 'name'.
#squash_actions = apk,apt,dnf,homebrew,pacman,pkgng,yum,zypper
```

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey.ps1:59]: I'm using an `ArrayList` for better performance and the [Add](https://msdn.microsoft.com/en-us/library/system.collections.arraylist.add.aspx) method returns the index of the added method which I need to redirect to `$null`.

The performance reasons are probably inconsequential at this number of elements but using `+=` on an actual array means PS needs to create a branch new array and copy each element + the new one to the new array and then destroy the old one which takes time. As a personal rule I try and use ArrayList where possible.

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 2 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey.ps1]: Thanks, will fix

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 3 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey.ps1]: `source` needs to be `null` so that `choco.exe` will use the configured sources in it's config. If we were to set the default to be the public repo then user's won't be able to control their sources through `win_chocolatey_source`. Also the URL is a bit different from the Chocolatey install vs `choco.exe install`, the installer is looking for a script at a particular location while `choco.exe install` should have either a name or a source, URL to a nuget feed, or a path to a local directory containing sources.

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 4 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey.ps1]: Yep that makes sense, will update it

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 5 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey.ps1]: Good point, will only fire when this was actually done.

File: lib/ansible/modules/windows/win_chocolatey.py (hunk 8 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey.py]: I had the proper YAML use earlier and through it best to show the alternative comma separated way but happy to put it back to a list to not encourage k=v syntax.

I personally don't think we want to add to the list of modules that support squashing considering non squash was the default behaviour and now people can choose either all and not be at the mercy of a global config.

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_chocolatey.ps1:59]: Hmm, surprising things... :-)

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_chocolatey.ps1]: Oh right. Must have been confused.

**Review by dagwieers — APPROVED** [review]

**Comment by jborean93** [comment]
Merging in so we can get people to test it out.

File: lib/ansible/modules/windows/win_chocolatey.ps1 (hunk 1 above)
Comment (Gianlu) [reply lib/ansible/modules/windows/win_chocolatey.ps1:59]: Hi,
may I suggest using 
`$arguments.Add("--debug") | out-null` instead of `$arguments.Add("--debug") > $null`.
I think it's more "Powershell Way".

Thanks

Comment (jborean93) [reply lib/ansible/modules/windows/win_chocolatey.ps1:59]: @Gianlu `| out-null` is a more expensive operation then casting to `[void]` or redirecting it directly to `$null` https://stackoverflow.com/questions/5260125/whats-the-better-cleaner-way-to-ignore-output-in-powershell. It’s not going to make a noticeable impact really but just something I recommend anyway in case you do deal with lots of these calls.

Pretty much the same reason why I’m using an ArrayList over a generic array.

Comment (Gianlu) [reply lib/ansible/modules/windows/win_chocolatey.ps1:59]: Ok, thanks for the explanation.
